### PR TITLE
Dev2017 flag and skip slow tests

### LINKED ISSUE #19864: Seperate out slow system tests from PR builds
Discover slow system tests, label them, and exclude them from PR builds.  Still run them for release and master builds

Slack Channel
dev2017_faster_builds

Slow tests spreadsheet
https://docs.google.com/spreadsheets/d/1uzAhnFM9Q8jMmrOzaZ32oBx_SDIdojKUgps9-EXpslo/edit#gid=0

Branch
https://github.com/mantidproject/mantid/tree/Dev2017_flag_and_skip_slow_tests

### PR DESCRIPTION
**Pull request by louisemccann**
A function `excludeInPullRequests` has been added to the base class for system tests, allowing for especially slow or unnecessary tests to be skipped for every PR build and only run nightly. Tests will be excluded if this function returns true.

When calling `systemtest` from the command line, all tests will be run by default. A flag `--exclude-in-pull-requests` has been added to exclude any systems test with the above function returning true.

A few of the slowest system tests have been updated to be skipped, see https://docs.google.com/spreadsheets/d/1uzAhnFM9Q8jMmrOzaZ32oBx_SDIdojKUgps9-EXpslo/edit#gid=0

**To test:**

Run a system test of your choice. 
Add the function
```
def excludeInPullRequests(self):
    return True
```
to the class of your system test.

run `systemtest -R yourtest` and it should be run.

run `systemtest -R yourtest --exclude-in-pull-requests` and it should be skipped.

Fixes #19864

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Testing/SystemTests/lib/systemtests/stresstesting.py [hunk 1]
```diff
@@ -563,23 +577,28 @@ def start(self, script):
 #########################################################################
 class TestScript(object):
 
-    def __init__(self, test_dir, module_name, test_cls_name):
+    def __init__(self, test_dir, module_name, test_cls_name, exinpr):
```
Comment (peterfpeterson) [inline Testing/SystemTests/lib/systemtests/stresstesting.py]: This is a horrible variable name. How about `exclude_in_pr_builds` or `excludeInPullRequests`?

File: Testing/SystemTests/lib/systemtests/stresstesting.py [hunk 2]
```diff
@@ -563,23 +577,28 @@ def start(self, script):
...
         self._test_cls_name = test_cls_name
+        self._exinpr = not exinpr
 
     def asString(self):
         code = '''import sys
 sys.path.append('{0}')
 sys.path.append('{1}')
 from {2} import {3}
 systest = {3}()
+def returnFalse():
```
Comment (peterfpeterson) [inline Testing/SystemTests/lib/systemtests/stresstesting.py]: This is a great case for using a [lambda function](http://www.diveintopython.net/power_of_introspection/lambda_functions.html).
```python
if {5}:
    systest.excludeInPullRequests = lambda : False
```

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Code looks good other than a few small changes.

File: Testing/SystemTests/scripts/runSystemTests.py [hunk 3]
```diff
@@ -52,6 +52,8 @@
                   help="A directory to use for the Mantid save path")
 parser.add_option("", "--archivesearch", dest="archivesearch", action="store_true",
                   help="Turn on archive search for file finder.")
+parser.add_option("", "--exclude-in-pull-requests", dest="exinpr",action="store_true",
```
Comment (peterfpeterson) [inline Testing/SystemTests/scripts/runSystemTests.py]: Better name here too

File: Testing/SystemTests/scripts/runSystemTests.py [hunk 4]
```diff
@@ -83,7 +85,7 @@
 runner = stresstesting.TestRunner(executable=options.executable, exec_args=execargs, escape_quotes=True)
 reporter = stresstesting.XmlResultReporter(showSkipped=options.showskipped)
 mgr = stresstesting.TestManager(mtdconf.testDir, runner, output = [reporter],
-                                testsInclude=options.testsInclude, testsExclude=options.testsExclude)
+                                testsInclude=options.testsInclude, testsExclude=options.testsExclude, exinpr=options.exinpr)
```
Comment (peterfpeterson) [inline Testing/SystemTests/scripts/runSystemTests.py]: Better name

**Review by peterfpeterson — CHANGES_REQUESTED** [review]

**Review by peterfpeterson — APPROVED** [review]
This works as advertised. Once the builds pass it can be merged.

**Comment by martyngigg** [comment]
Good stuff. I assume we need to add the flag to the `pull_requests-rhel7` configuration?

**Comment by peterfpeterson** [comment]
The change to the build script is done in #19913

**Pull request merged**